contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,268 @@
+     * 1- Based on the LP token to be deposited, find out the Onsen Pool Id
+     * 2- Deposit LP token into onsen, if deposit is succesfull we should get in return sushi tokens
+     * 2- Stake the Sushi rewards tokens returned from deposit LP tokens and as rewards form onsen LP stake, we should get in return xSushi tokens
+     * 3- Keep the xSushi on the allocator contract for easily deallocation
+     * 4- Calculate gains/loss based on the LP balance on the Onsen Pool.
+     */
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        uint256 balance = LPtoken.balanceOf(address(this));
+
+        /// Deposit LP token into onsen, if deposit succesfull this address should have in return sushi tokens
+        if (balance > 0) {
+            /// Find out if there is a Onsen Pool for the LPToken
+            (bool onsenLPFound, uint256 onsenPoolId) = _findPoolByLP(address(LPtoken));
+
+            if (onsenLPFound) {
+                /// Approve and deposit balance into onsen Pool
+                LPtoken.approve(masterChef, balance);
+                IMasterChef(masterChef).deposit(onsenPoolId, balance, address(this));
+
+                /// Stake the sushi tokens
+                _stakeSushi();
+
+                ///Calculate gains/loss
+                /// Retrieve current balance for pool and address
+                UserInfo memory currentUserInfo = IMasterChef(masterChef).userInfo(onsenPoolId, address(this));
+
+                uint256 last = extender.getAllocatorAllocated(id) + extender.getAllocatorPerformance(id).gain;
+
+                if (currentUserInfo.amount >= last) {
+                    gain = uint128(currentUserInfo.amount - last);
+                } else {
+                    loss = uint128(last - currentUserInfo.amount);
+                }
+            } else {
+                /// If no Onsen pool was found for the LP token then return LP token to the treasury.
+                LPtoken.safeTransfer(treasury, balance);
+            }
+        }
+    }
+
+    function deallocate(uint256[] memory amounts) public override onlyGuardian {
+        for (uint256 i; i < amounts.length; i++) {
+            uint256 amount = amounts[i];
+            if (amount > 0) {
+                /// Get the LPToken
+                IERC20 token = _tokens[i];
+
+                /// Get the onsen pool Id
+                (bool onsenLPFound, uint256 onsenPoolId) = _findPoolByLP(address(token));
+
+                if (onsenLPFound) {
+                    /// Withdraw LP Token from Onsen Pool
+                    IMasterChef(masterChef).withdrawAndHarvest(onsenPoolId, amount, address(this));
+                }
+            }
+        }
+    }
+
+    function _deactivate(bool panic) internal override {
+        /// Get amounts by LP to unstake and then deallocate (unstake) each token.
+        _deallocateAll();
```

Comment:
we can actually remove it in this case and just set all of it under panic

Reply:
_deactivate function updated.